docs/proposals/user-namespace.md

```diff
@@ -0,0 +1,226 @@
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# User Namespaces
+
+## Motivation
+
+Running a container as root is a common practice for many
+image authors. However, many clusters operate under a more strict
+security profile that prevents usage of the root UID. This means that
+either admins must allow exceptions for root access to run many of
+the images available or image authors must be aware of the deployment
+environment in order to develop a container that works well under
+the increased security constraints. This frustrates end-users that
+want to run the same images they run on their host as they cannot
+run those images on the cluster.
+
+Docker 1.10 supports [user namespace remapping](https://docs.docker.com/v1.10/engine/reference/commandline/daemon/#daemon-user-namespace-options)
+in the docker daemon.  This allows you to run as the root user in a
+container while running as a non-root user in the host.
+
+Additionally, the `HostConfig` for a container now contains a parameter
+that lets you escape the private user namespace and indicate that you
+would like to [use the host's user namespace](https://github.com/docker/engine-api/blob/master/types/container/host_config.go#L310).
+By using this setting, administrators may run containers outside of the user
+namespace while still requiring other containers to run as a remapped
+user.
+
+## Current State
+
+In Docker, you may provide the `--userns-remap` option to the daemon
+to enable remapping.  This option will read the `/etc/subuid` and
+the `/etc/subgid` files to determine [how to map UIDs and GIDs
+in the container](https://docs.docker.com/engine/reference/commandline/dockerd/#detailed-information-on-subuidsubgid-ranges).
+
+Currently there are no configuration options available to make
+Kubernetes aware of a container runtime running a remapped user
+environment.
+
+If the Docker daemon is launched with the remap option in a Kubernetes
+cluster today and a pod author would like to run a container that uses
+the host user namespace there is not currently a way to express the
+configuration.
+
+### Known Issues
+
+These are the known issues (as of writing) that will be hit when
+enabling user namespace remapping in a Kubernetes cluster.
+
+1. Negative oomScoreAdj values fail.  In this case the pause container
+cannot be started and shows an error `write /proc/self/oom_score_adj: permission denied`
+in the daemon log.  Failing in docker 1.10, working in 1.12
+1. Cannot share the host or a container's network namespace when user
+namespaces are enabled.  [Fixed in 1.11](https://github.com/docker/docker/pull/21383).
+1. Permission denied mounting.  Occurs due to ownership permissions
+in the `/var/lib/kubelet/pods` directory when trying to mount the
+termination logs and secrets.  A workaround (not a solution) is to
+change the remapped GID to the root group.
+
+## Design
+
+### Constraints
+
+* If containers share any namespace they must share the user namespace.
```

Comment:
Each non-user ns has an owning user ns. If net is set to host, then the owning user ns for net will be host user ns and by extension for ipc as well. So, statement 2 is valid but could be reworded better.